### PR TITLE
Refactor doctest

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Main changes:

  - run doctest in multiple independent jobs
    - motivation: to avoid GPU OOM (or other environment states) affect subsequent tests
    - this is done by using a map between directories and their files
    - The files in `docs/source/en/model_doc` or `docs/source/en/tasks` are **NOT** grouped together with other files in the
same directory: the objective is to run doctest against them in independent GitHub Actions jobs. 
  - Use the same trick as in #28773 to allow GitHub Actions to generate more than 256 jobs in matrix
    - (as now we have many more (smaller) jobs instead of a single big job)

A file `doc_test_results` is saved and uploaded as artifact so we can always have a way to check.

A (example) run page: see [here](https://github.com/huggingface/transformers/actions/runs/8659176586)
A full run page: see [here](https://github.com/huggingface/transformers/actions/runs/8659540805)

### CODE REVIEW

File: .github/workflows/doctest_job.yml
Comment (ydshieh) [inline .github/workflows/doctest_job.yml:1]: A new workflow file that contain the doctest job definition (logic copied from existing file `doctests.yml`

File: .github/workflows/doctests.yml
Comment (ydshieh) [inline .github/workflows/doctests.yml:1]: move the doctest job definition to the new workflow file. Now, this workflow only get the list of files to run doctest and how we run it (using GitHub Actions matrix)

File: utils/notification_service_doc_tests.py
Comment (ydshieh) [inline utils/notification_service_doc_tests.py:1]: Nothing really changed: previously, we only have one artifact but now we have multiple artifacts. The change here is just to taking into account of this fact to aggregate the information (test results) across multiple reports.

File: utils/split_doctest_jobs.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+To get a map (dictionary) between directory (or file) paths and the corresponding files
+```bash
+python utils/split_doctest_jobs.py
+```
+or to get a list of lists of directory (or file) paths
+```bash
+python utils/split_doctest_jobs.py --only_return_keys --num_splits 4
+```
+(this is used to allow GitHub Actions to generate more than 256 jobs using matrix)
+"""
```
Comment (ydshieh) [inline utils/split_doctest_jobs.py:32]: check here

File: utils/tests_fetcher.py [hunk 2]
```diff
@@ -502,7 +502,10 @@ def get_all_doctest_files() -> List[str]:
     """
     py_files = [str(x.relative_to(PATH_TO_REPO)) for x in PATH_TO_REPO.glob("**/*.py")]
     md_files = [str(x.relative_to(PATH_TO_REPO)) for x in PATH_TO_REPO.glob("**/*.md")]
+
     test_files_to_run = py_files + md_files
+    # change to use "/" as path separator
+    test_files_to_run = ["/".join(Path(x).parts) for x in test_files_to_run]
```
Comment (ydshieh) [inline utils/tests_fetcher.py:508]: make everything using `/`: so it will works on `Windows` too

**Review by LysandreJik — APPROVED** [review]
Looks clean! Thanks @ydshieh!

**Comment by ydshieh** [comment]
Merge now :-)

**Pull request merged**